x-pack/plugins/alerting/server/lib/get_alerts.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { millisToNanos } from '@kbn/event-log-plugin/server';
+import { Alert } from '../alert';
+import { AlertInstanceState, AlertInstanceContext } from '../types';
+
+interface GetAlertsResult<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  newAlerts: Record<string, Alert<State, Context, ActionGroupIds>>;
+  activeAlerts: Record<string, Alert<State, Context, ActionGroupIds>>;
+  recoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupId>>;
+}
+
+export function getAlerts<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>(
+  alerts: Record<string, Alert<State, Context>>,
+  originalAlerts: Record<string, Alert<State, Context>>
+): GetAlertsResult<State, Context, ActionGroupIds, RecoveryActionGroupId> {
+  const originalAlertIds = new Set(Object.keys(originalAlerts));
+
+  const currentTime = new Date().toISOString();
+  const newAlerts: Record<string, Alert<State, Context, ActionGroupIds>> = {};
+  const activeAlerts: Record<string, Alert<State, Context, ActionGroupIds>> = {};
+  const recoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupId>> = {};
+
+  for (const id in alerts) {
+    if (alerts.hasOwnProperty(id)) {
+      // alerts with scheduled actions are considered "active"
+      if (alerts[id].hasScheduledActions()) {
+        activeAlerts[id] = alerts[id];
+
+        // if this alert did not exist in previous run, it is considered "new"
+        if (!originalAlertIds.has(id)) {
+          newAlerts[id] = alerts[id];
+
+          // Inject start time into alert state for new alerts
+          const state = newAlerts[id].getState();
+          newAlerts[id].replaceState({ ...state, start: currentTime, duration: '0' });
```

Comment:
This feels like a side effect, i expect getAlert to return alerts as groups, but it also modifies them.

Reply:
How about naming the function `getAlertsWithDuration`?